### PR TITLE
[FEATURE] Add support for TYPO3 8.7 (#139)

### PR DESCRIPTION
**Pull request by srotsch**

### CODE REVIEW

File: class.tx_crawler_lib.php [hunk 1]
```diff
@@ -698,7 +756,7 @@ public function expandParameters($paramArray, $pid)
                 foreach ($parts as $pV) {
 
                         // Look for integer range: (fx. 1-34 or -40--30 // reads minus 40 to minus 30)
-                    if (preg_match('/^(-?[0-9]+)\s*-\s*(-?[0-9]+)$/', trim($pV), $reg)) {	// Integer range:
+                    if (preg_match('/^(-?[0-9]+)\s*-\s*(-?[0-9]+)$/', trim($pV), $reg)) {    // Integer range:
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces before comment

File: class.tx_crawler_lib.php [hunk 2]
```diff
@@ -708,7 +766,7 @@ public function expandParameters($paramArray, $pid)
                         }
 
                         // Traverse range, add values:
-                        $runAwayBrake = 1000;	// Limit to size of range!
+                        $runAwayBrake = 1000;    // Limit to size of range!
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces before comment

File: class.tx_crawler_lib.php [hunk 3]
```diff
@@ -760,7 +818,7 @@ public function expandParameters($paramArray, $pid)
                                 }
                             }
                         }
-                    } else {	// Just add value:
+                    } else {    // Just add value:
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces before comment

File: class.tx_crawler_lib.php [hunk 4]
```diff
@@ -873,14 +932,15 @@ public function getLogEntriesForPageId($id, $filter = '', $doFlush = false, $doF
...
-     * @param	integer		Set ID for which to look up log entries.
-     * @param	string		Filter: "all" => all entries, "pending" => all that is not yet run, "finished" => all complete ones
-     * @param	boolean		If TRUE, then entries selected at DELETED(!) instead of selected!
-     * @param	integer		Limit the amount of entires per page default is 10
-     * @return	array
+     * @param    integer        Set ID for which to look up log entries.
+     * @param    string        Filter: "all" => all entries, "pending" => all that is not yet run, "finished" => all complete ones
+     * @param    boolean        If TRUE, then entries selected at DELETED(!) instead of selected!
+     * @param    integer        Limit the amount of entires per page default is 10
+     * @return    array
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 5]
```diff
@@ -911,7 +971,7 @@ public function getLogEntriesForSetId($set_id, $filter = '', $doFlush = false, $
     /**
      * Removes queue entires
      *
-     * @param $where	SQL related filter for the entries which should be removed
+     * @param $where    SQL related filter for the entries which should be removed
      * @return void
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 6]
```diff
@@ -931,12 +991,12 @@ protected function flushQueue($where = '')
...
-     * @param	string		Call back object reference, eg. 'EXT:indexed_search/class.crawler.php:&tx_indexedsearch_crawler'
-     * @param	integer		Page ID to attach it to
-     * @param	integer		Time at which to activate
-     * @return	void
+     * @param    integer        Set ID
+     * @param    array        Parameters to pass to call back function
+     * @param    string        Call back object reference, eg. 'EXT:indexed_search/class.crawler.php:&tx_indexedsearch_crawler'
+     * @param    integer        Page ID to attach it to
+     * @param    integer        Time at which to activate
+     * @return    void
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 7]
```diff
@@ -967,13 +1027,13 @@ public function addQueueEntry_callBack($setId, $params, $callBack, $page_id = 0,
...
-     * @param 	string		(optional) configuration hash
-     * @param 	bool		(optional) skip inner duplication check
-     * @return	bool		true if the url was added, false if it already existed
+     * @param    integer        Page ID
+     * @param    string        Complete URL
+     * @param    array        Sub configuration array (from TS config)
+     * @param    integer        Scheduled-time
+     * @param     string        (optional) configuration hash
+     * @param     bool        (optional) skip inner duplication check
+     * @return    bool        true if the url was added, false if it already existed
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 8]
```diff
@@ -1200,8 +1260,8 @@ public function readUrl($queueId, $force = false)
     /**
      * Read URL for not-yet-inserted log-entry
      *
-     * @param	integer		Queue field array,
-     * @return	string
+     * @param    integer        Queue field array,
+     * @return    string
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 9]
```diff
@@ -1230,16 +1290,16 @@ public function readUrlFromArray($field_array)
     /**
      * Read URL for a queue record
      *
-     * @param	array		Queue record
-     * @return	string		Result output.
+     * @param    array        Queue record
+     * @return    string        Result output.
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 10]
```diff
@@ -1230,16 +1290,16 @@ public function readUrlFromArray($field_array)
...
      */
     public function readUrl_exec($queueRec)
     {
         // Decode parameters:
         $parameters = unserialize($queueRec['parameters']);
         $result = 'ERROR';
         if (is_array($parameters)) {
-            if ($parameters['_CALLBACKOBJ']) {	// Calling object:
+            if ($parameters['_CALLBACKOBJ']) {    // Calling object:
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces before comment

File: class.tx_crawler_lib.php [hunk 11]
```diff
@@ -1248,7 +1308,7 @@ public function readUrl_exec($queueRec)
                 } else {
                     $result = ['content' => 'No object: '.$objRef];
                 }
-            } else {	// Regular FE request:
+            } else {    // Regular FE request:
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces before comment

File: class.tx_crawler_lib.php [hunk 12]
```diff
@@ -1478,12 +1539,12 @@ public function buildRequestHeaderArray(array $url, $crawlerId)
...
      * Check if the submitted HTTP-Header contains a redirect location and built new crawler-url
      *
-     * @param	array		HTTP Header
-     * @param	string		HTTP Auth. User
-     * @param	string		HTTP Auth. Password
-     * @return	string		URL from redirection
+     * @param    array        HTTP Header
+     * @param    string        HTTP Auth. User
+     * @param    string        HTTP Auth. Password
+     * @return    string        URL from redirection
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 13]
```diff
@@ -1527,9 +1588,9 @@ public function getRequestUrlFrom302Header($headers, $user = '', $pass = '')
      * Initialization hook (called after database connection)
      * Takes the "HTTP_X_T3CRAWLER" header and looks up queue record and verifies if the session comes from the system (by comparing hashes)
      *
-     * @param	array		Parameters from frontend
-     * @param	object		TSFE object (reference under PHP5)
-     * @return	void
+     * @param    array        Parameters from frontend
+     * @param    object        TSFE object (reference under PHP5)
+     * @return    void
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 14]
```diff
@@ -1557,15 +1618,15 @@ public function fe_init(&$params, $ref)
...
-     * @return	string		HTML code
+     * @param    integer        Root page id to start from.
+     * @param    integer        Depth of tree, 0=only id-page, 1= on sublevel, 99 = infinite
+     * @param    integer        Unix Time when the URL is timed to be visited when put in queue
+     * @param    integer        Number of requests per minute (creates the interleave between requests)
+     * @param    boolean        If set, submits the URLs to queue in database (real crawling)
+     * @param    boolean        If set (and submitcrawlUrls is false) will fill $downloadUrls with entries)
+     * @param    array        Array of processing instructions
+     * @param    array        Array of configuration keys
+     * @return    string        HTML code
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 15]
```diff
@@ -1603,9 +1664,9 @@ public function getPageTreeAndUrls(
 
         // Set root row:
         $tree->tree[] = [
-            'row' => $pageinfo,
+                'row' => $pageinfo,
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 16]
```diff
@@ -1603,9 +1664,9 @@ public function getPageTreeAndUrls(
...
             'HTML' => \AOE\Crawler\Utility\IconUtility::getIconForRecord('pages', $pageinfo)
-        ];
+            ];
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 17]
```diff
@@ -1710,9 +1771,9 @@ public function expandExcludeString($excludeString)
      * Create the rows for display of the page tree
      * For each page a number of rows are shown displaying GET variable configuration
      *
-     * @param	array		Page row
-     * @param	string		Page icon and title for row
-     * @return	string		HTML <tr> content (one or more)
+     * @param    array        Page row
+     * @param    string        Page icon and title for row
+     * @return    string        HTML <tr> content (one or more)
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 18]
```diff
@@ -1722,7 +1783,7 @@ public function drawURLs_addRowsForPage(array $pageRow, $pageTitleAndIcon)
         $configurations = $this->getUrlsForPageRow($pageRow, $skipMessage);
 
         if (count($this->incomingConfigurationSelection) > 0) {
-            // 	remove configuration that does not match the current selection
+            //     remove configuration that does not match the current selection
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces in comment

File: class.tx_crawler_lib.php [hunk 19]
```diff
@@ -1765,13 +1826,13 @@ public function drawURLs_addRowsForPage(array $pageRow, $pageTitleAndIcon)
...
+                            <tr>
+                                <td class="bgColor4-20">'.htmlspecialchars('&'.$gVar.'=').'<br/>'.
                                                 '('.count($gVal).')'.
                                                 '</td>
-								<td class="bgColor4" nowrap="nowrap">'.nl2br(htmlspecialchars(implode(chr(10), $gVal))).'</td>
-							</tr>
-						';
+                                <td class="bgColor4" nowrap="nowrap">'.nl2br(htmlspecialchars(implode(chr(10), $gVal))).'</td>
+                            </tr>
+                        ';
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 20]
```diff
@@ -1792,21 +1853,21 @@ public function drawURLs_addRowsForPage(array $pageRow, $pageTitleAndIcon)
...
                 } else {
                     $content .= '<tr class="bgColor'.($c % 2 ? '-20':'-10') . '">
-							'.$titleClm.'
-							<td>'.htmlspecialchars($confKey).'</td>
-							<td colspan="5"><em>No entries</em> (Page is excluded in this configuration)</td>
-						</tr>';
+                            '.$titleClm.'
+                            <td>'.htmlspecialchars($confKey).'</td>
+                            <td colspan="5"><em>No entries</em> (Page is excluded in this configuration)</td>
+                        </tr>';
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 21]
```diff
@@ -1816,10 +1877,10 @@ public function drawURLs_addRowsForPage(array $pageRow, $pageTitleAndIcon)
...
             // Compile row:
             $content .= '
-				<tr class="bgColor-20" style="border-bottom: 1px solid black;">
-					<td>'.$pageTitleAndIcon.'</td>
-					<td colspan="6"><em>No entries</em>'.$message.'</td>
-				</tr>';
+                <tr class="bgColor-20" style="border-bottom: 1px solid black;">
+                    <td>'.$pageTitleAndIcon.'</td>
+                    <td colspan="6"><em>No entries</em>'.$message.'</td>
+                </tr>';
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 22]
```diff
@@ -1835,8 +1896,8 @@ public function getUnprocessedItemsCount()
                     'count(*) as num',
                     'tx_crawler_queue',
                     'exec_time=0
-					AND process_scheduled= 0
-					AND scheduled<='.$this->getCurrentTime()
+                    AND process_scheduled= 0
+                    AND scheduled<='.$this->getCurrentTime()
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 23]
```diff
@@ -1853,7 +1914,7 @@ public function getUnprocessedItemsCount()
      * Main function for running from Command Line PHP script (cron job)
      * See ext/crawler/cli/crawler_cli.phpsh for details
      *
-     * @return	int number of remaining items or false if error
+     * @return    int number of remaining items or false if error
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 24]
```diff
@@ -1899,7 +1960,7 @@ public function CLI_main()
     /**
      * Function executed by crawler_im.php cli script.
      *
-     * @return	void
+     * @return    void
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 25]
```diff
@@ -1950,7 +2011,7 @@ public function CLI_main_im()
             tx_crawler_domain_events_dispatcher::getInstance()->post(
                 'invokeQueueChange',
                 $this->setID,
-                [	'reason' => $reason ]
+                [    'reason' => $reason ]
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 26]
```diff
@@ -2087,8 +2148,8 @@ public function CLI_run($countInARun, $sleepTime, $sleepAfterFinish)
             'qid,scheduled',
             'tx_crawler_queue',
             'exec_time=0
-				AND process_scheduled= 0
-				AND scheduled<='.$this->getCurrentTime(),
+                AND process_scheduled= 0
+                AND scheduled<='.$this->getCurrentTime(),
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 27]
```diff
@@ -2137,7 +2198,7 @@ public function CLI_run($countInARun, $sleepTime, $sleepAfterFinish)
                 $result |= $this->readUrl($r['qid']);
 
                 $counter++;
-                usleep(intval($sleepTime));	// Just to relax the system
+                usleep(intval($sleepTime));    // Just to relax the system
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 28]
```diff
@@ -2150,7 +2211,7 @@ public function CLI_run($countInARun, $sleepTime, $sleepAfterFinish)
 
                     //TODO might need an additional returncode
                     $result |= self::CLI_STATUS_ABORTED;
-                    break;		//possible timeout
+                    break;        //possible timeout
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 29]
```diff
@@ -2172,7 +2233,7 @@ public function CLI_run($countInARun, $sleepTime, $sleepAfterFinish)
     /**
      * Activate hooks
      *
-     * @return	void
+     * @return    void
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: class.tx_crawler_lib.php [hunk 30]
```diff
@@ -2288,11 +2349,11 @@ public function CLI_releaseProcesses($releaseIds, $withinLock = false)
...
-			AND NOT EXISTS (
-				SELECT * FROM tx_crawler_queue
-				WHERE tx_crawler_queue.process_id = tx_crawler_process.process_id
-				AND tx_crawler_queue.exec_time = 0
-			)',
+            AND NOT EXISTS (
+                SELECT * FROM tx_crawler_queue
+                WHERE tx_crawler_queue.process_id = tx_crawler_process.process_id
+                AND tx_crawler_queue.exec_time = 0
+            )',
```
Comment (srotsch) [inline class.tx_crawler_lib.php]: Check whitespaces

File: domain/process/class.tx_crawler_domain_process_manager.php [hunk 31]
```diff
@@ -102,7 +105,7 @@ public function multiProcess($timeout)
             }
             if ($currentPendingItems == 0) {
                 if ($this->verbose) {
-                    echo 'Finished...' . chr(10);
+                    echo 'Finished...'.chr(10);
```
Comment (srotsch) [inline domain/process/class.tx_crawler_domain_process_manager.php]: Missing whitespaces around concatenation operator

File: domain/process/class.tx_crawler_domain_process_manager.php [hunk 32]
```diff
@@ -114,16 +117,16 @@ public function multiProcess($timeout)
             }
             sleep(1);
             if ($nextTimeOut < time()) {
-                $timedOutProcesses = $this->processRepository->findAll('', 'DESC', null, 0, 'ttl >' . $nextTimeOut);
+                $timedOutProcesses = $this->processRepository->findAll('', 'DESC', null, 0, 'ttl >'.$nextTimeOut);
```
Comment (srotsch) [inline domain/process/class.tx_crawler_domain_process_manager.php]: Missing whitespaces around concatenation operator

File: domain/process/class.tx_crawler_domain_process_manager.php [hunk 33]
```diff
@@ -114,16 +117,16 @@ public function multiProcess($timeout)
...
                 $nextTimeOut = time() + $this->timeToLive;
                 if ($this->verbose) {
-                    echo 'Cleanup' . implode(',', $timedOutProcesses->getProcessIds()) . chr(10);
+                    echo 'Cleanup'.implode(',', $timedOutProcesses->getProcessIds()).chr(10);
```
Comment (srotsch) [inline domain/process/class.tx_crawler_domain_process_manager.php]: Missing whitespaces around concatenation operator

File: domain/process/class.tx_crawler_domain_process_manager.php [hunk 34]
```diff
@@ -114,16 +117,16 @@ public function multiProcess($timeout)
...
                 }
                 $this->crawlerObj->CLI_releaseProcesses($timedOutProcesses->getProcessIds(), true);
             }
         }
         if ($currentPendingItems > 0 && $this->verbose) {
-            echo 'Stop with timeout' . chr(10);
+            echo 'Stop with timeout'.chr(10);
```
Comment (srotsch) [inline domain/process/class.tx_crawler_domain_process_manager.php]: Missing whitespaces around concatenation operator

File: ext_localconf.php [hunk 35]
```diff
@@ -28,6 +29,6 @@
         'os' => '',
         'exec' => '',
 
-        'className' => 'tx_crawler_auth',
+        'className' => 'AOE\\Crawler\\Service\\AuthenticationService',
```
Comment (srotsch) [inline ext_localconf.php]: ::class

File: class.tx_crawler_lib.php (hunk 19 above)
Comment (tomasnorre) [reply class.tx_crawler_lib.php]: The php-cs-fixer should already have been used.. Let's check again..

**Comment by tmotyl** [comment]
Great to see, thanks a lot guys!

**Pull request merged**